gateway/handlers/createhandler.go

```diff
@@ -167,3 +179,65 @@ func userPasswordFromBasicAuth(basicAuthB64 string) (string, string, error) {
 	}
 	return cs[:s], cs[s+1:], nil
 }
+
+func makeSecretsArray(c *client.Client, secretNames []string) ([]*swarm.SecretReference, error) {
+	values := []*swarm.SecretReference{}
+
+	if len(secretNames) == 0 {
+		return values, nil
+	}
+
+	secretOpts := new(opts.SecretOpt)
+	for _, secret := range secretNames {
+		if err := secretOpts.Set(secret); err != nil {
+			return nil, err
+		}
+	}
+
+	requestedSecrets := make(map[string]bool)
+	ctx := context.Background()
+
+	// query the Swarm for the requested secret ids, these are required to complete
+	// the spec
+	args := filters.NewArgs()
+	for _, opt := range secretOpts.Value() {
+		args.Add("name", opt.SecretName)
+	}
+
+	secrets, err := c.SecretList(ctx, types.SecretListOptions{
+		Filters: args,
+	})
+	if err != nil {
```

Comment:
Should we handle this and make a 404 go back to the client?

Reply:
That is the responsibility of the caller, which we currently do 

```go
		secrets, err := makeSecretsArray(c, request.Secrets)
		if err != nil {
			log.Println(err)
			w.WriteHeader(http.StatusBadRequest)
			w.Write([]byte("Deployment error: " + err.Error()))
			return
		}
```

Reply:
What happens if you do this out of order:

* Create function (asking for secret1)
* Create secret1 afterwards

Does the function eventually pick up the secret?

Reply:
I am pretty sure the secret must exist first, for example if you take the command from [the walk through on docker secrets](https://docs.docker.com/engine/swarm/secrets/#simple-example-get-started-with-secrets) 

```sh
docker service  create --name redis --secret my_secret_data redis:alpine
```
without having first create the `my_secret_data`   it replies with 

```
secret not found: my_secret_data
```

So, secrets must be created before functions, otherwise docker will throw an error.